// @sudoku/stores/gameStore.js

import { writable } from 'svelte/store';
import { decodeSencode } from '@sudoku/sencode';
import { generateSudoku } from '@sudoku/sudoku';
import { Sudoku } from '../domain/Sudoku.js';
import { Game } from '../domain/Game.js';

function createEmptyGrid() {
  return [
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
  ];
}

function createEmptyState() {
  return {
    initialGrid: createEmptyGrid(),
    grid: createEmptyGrid(),
    invalidCells: [],
    won: false,
    canUndo: false,
    canRedo: false,
    ready: false,
  };
}

function createGameStore() {
  const { subscribe, set } = writable(createEmptyState());
  let game = null;

  function sync() {
    if (!game) {
      set(createEmptyState());
      return;
    }

    set({
      ...game.getViewData(),
      ready: true,
    });
  }

  return {
    subscribe,

    startNew(difficulty = 'easy') {
      const puzzle = generateSudoku(difficulty);
      game = new Game(new Sudoku(puzzle));
      sync();
    },

    startCustom(source) {
      const puzzle = typeof source === 'string' ? decodeSencode(source) : source;
      game = new Game(new Sudoku(puzzle));
      sync();
    },

    guess(move) {
      if (!game) return false;
      const ok = game.guess(move);
      if (ok) sync();
      return ok;
    },

    applyHint(pos) {
      if (!game) return false;
      const ok = game.applyHint(pos);
      if (ok) sync();
      return ok;
    },

    undo() {
      if (!game) return false;
      const ok = game.undo();
      if (ok) sync();
      return ok;
    },

    redo() {
      if (!game) return false;
      const ok = game.redo();
      if (ok) sync();
      return ok;
    },

    loadFromJSON(json) {
      game = Game.fromJSON(json);
      sync();
    },

    toJSON() {
      return game ? game.toJSON() : null;
    }
  };
}

export const gameStore = createGameStore();
